Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp

```diff
@@ -147,24 +152,30 @@ MediaStreamTrackProcessor::VideoFrameObserver::~VideoFrameObserver()
 
 RefPtr<WebCodecsVideoFrame> MediaStreamTrackProcessor::VideoFrameObserver::takeVideoFrame(ScriptExecutionContext& context)
 {
-    Locker lock(m_videoFrameLock);
-    if (!m_videoFrame)
-        return nullptr;
+    RefPtr<VideoFrame> videoFrame;
+    {
+        Locker lock(m_videoFramesLock);
+        if (m_videoFrames.isEmpty())
+            return nullptr;
+
+        videoFrame = m_videoFrames.takeFirst();
+    }
 
     WebCodecsVideoFrame::BufferInit init;
-    init.codedWidth = m_videoFrame->presentationSize().width();
-    init.codedHeight = m_videoFrame->presentationSize().height();
-    init.colorSpace = m_videoFrame->colorSpace();
+    init.codedWidth = videoFrame->presentationSize().width();
+    init.codedHeight = videoFrame->presentationSize().height();
+    init.colorSpace = videoFrame->colorSpace();
 
-    return WebCodecsVideoFrame::create(context, m_videoFrame.releaseNonNull(), WTFMove(init));
+    return WebCodecsVideoFrame::create(context, videoFrame.releaseNonNull(), WTFMove(init));
 }
 
-void MediaStreamTrackProcessor::VideoFrameObserver::videoFrameAvailable(VideoFrame& frame, VideoFrameTimeMetadata metadata)
+void MediaStreamTrackProcessor::VideoFrameObserver::videoFrameAvailable(VideoFrame& frame, VideoFrameTimeMetadata)
 {
     {
-        Locker lock(m_videoFrameLock);
-        m_videoFrame = &frame;
-        m_metadata = metadata;
+        Locker lock(m_videoFramesLock);
+        m_videoFrames.append(frame);
+        if (m_videoFrames.size() > m_maxVideoFramesCount)
+            m_videoFrames.takeFirst();
```

Comment:
It might be worth logging (debug level only?) when a frame is dropped.

Reply:
Sure, I will add some logging for that.